feat(widget): validate partner fee

shoom3301 opened the pull request:
# Summary

1. Added controls in the widget configurator for partner fee percent and recipient address
2. Added validation on the app side for the fee percent and address
3. When there are some errors on the widget parameters then block the whole app and display the errors. @fairlighteth we need your power here 🙏 

<img width="295" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/c4608777-c4b1-4666-b8f0-b3d177344e33">

<img width="1165" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/012ebced-ec0f-454d-bf5a-96a6d9eb5c95">

  # To Test

1. Set a partner fee bellow 0%
- [ ] "Partner fee can not be less than 0%!" error should be displayed
2. Set a partner fee above 1%
- [ ] "Partner fee can not be more than 1%" error should be displayed
3. When the recipient is specified and not a valid address
- [ ] Then corresponding error should be displayed

# Questions

1. Should a recipient address be always specified along with the fee?

shoom3301 commented on apps/cowswap-frontend/src/modules/injectedWidget/consts.ts:
```diff
@@ -1 +1,3 @@
 export const COW_SWAP_WIDGET_EVENT_KEY = 'cowSwapWidget'
+
+export const PARTNER_FEE_MAX_BPS = 5000 // 50%
```
TODO: adjust the value

shoom3301 commented on apps/cowswap-frontend/src/modules/injectedWidget/pure/WidgetParamsErrorsScreen/index.tsx:
```diff
@@ -0,0 +1,30 @@
+  if (paramsWithErrors.length === 0) return null
+
+  return (
+    <Container>
```
TODO: add beauty here

anxolin reviewed (commented):
I see something strange in the values:

<img width="1530" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/3825236c-7d75-4c92-9ed4-f09d62fed87a">

anxolin reviewed (commented):
Should the error be expressed also in BPS instead of percentage? 
No more than 100 BPS

<img width="1288" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/cf6fcd44-aa46-4299-a170-a39d267b716e">

elena-zh commented:
Hey @shoom3301 , I've got a crash when I filled in 0.001 value 
![image](https://github.com/cowprotocol/cowswap/assets/70885163/57b5c6df-93ee-407c-999e-0426f4ba088c)

elena-zh commented:
As for the Recipient validation, can you please make the address symbols case insensitive?  
![image](https://github.com/cowprotocol/cowswap/assets/70885163/080a43a2-206c-4edb-ad4a-ebf5a2c222e5)


Not a top priority, but when an error is displayed, it would be great to show it's actual value (if changed)
![image](https://github.com/cowprotocol/cowswap/assets/70885163/bf82a64c-93ca-47c8-82c4-b46beca105ff)

Question: is this OK to allow users to fill in 0x0000000000000000000000000000000000000000 address?
![image](https://github.com/cowprotocol/cowswap/assets/70885163/458a1cc9-2a75-4c7e-a9fd-0d8f327e2681)

shoom3301 commented:
> I see something strange in the values:
> 
> <img alt="image" width="1530" src="https://private-user-images.githubusercontent.com/2352112/309265456-3825236c-7d75-4c92-9ed4-f09d62fed87a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk2MjM5NDAsIm5iZiI6MTcwOTYyMzY0MCwicGF0aCI6Ii8yMzUyMTEyLzMwOTI2NTQ1Ni0zODI1MjM2Yy03ZDc1LTRjOTItOWVkNC1mMDlkNjJmZWQ4N2EucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDVUMDcyNzIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDQ4NDY1Y2IwODIyYmM4ZjVkZjhkMzIzM2MyNmY1YzFmN2JkMzY1MDFmYzk4NzhkY2VjMGY0ODMxMGVmNGQzZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.uIyDA7It2ooHAj9zpA_u1kuFWDckpguYqsg74TetCdY">

@anxolin Sorry, 1% of 1000 should be 10.
1000 = 100%
   x     = 1%

 x = 1000 * 1 / 100
 x = 10

shoom3301 commented:
>Should the error be expressed also in BPS instead of percentage?
>No more than 100 BPS

@anxolin fixed it, thanks!

shoom3301 commented:
@elena-zh thank you for the feedback!
1. I think recipient always myst be set if partner fee is specified, so I added "Partner fee recipient must be set!" error
2. We don't support ENS here
3. Disabled a possibility to paste decimals in the BPS input
4. Recipient address validation must be case sensitive
5. Also fixed the error displaying (change it reactively)
6. 0x0000...0000 is a valid recipient address, I've even set it as default

elena-zh commented:
Hey @shoom3301 , great, thank you!
The only one nitpick is that I'd not to show recipient error message once fee is set to 0
![image](https://github.com/cowprotocol/cowswap/assets/70885163/3ba31209-ba70-4ca6-a9fd-0a10139d839e)

shoom3301 commented:
@elena-zh fixed

elena-zh reviewed (approved):
Great, thank you!

anxolin commented:
Closes #3977 #3975

alfetopito commented on apps/widget-configurator/src/app/configurator/controls/PartnerFeeControl.tsx:
```diff
@@ -0,0 +1,40 @@
+  const [feeBps, setFeeBps] = feeBpsState
+  const [recipient, setRecipient] = recipientState
+
+  const onFeeBpsChange = (e: ChangeEvent<HTMLInputElement>) => {
+    const value = Math.ceil(Number(e.target.value || 0))
+
+    setFeeBps(value)
+  }
```
useCallback?

shoom3301 replied:
Fixed

anxolin commented:
> should Recipient field should be mandaroty when a user fills in the Partner fee field?

Yes! In the configurator I don't care too much, we could even potentially hide it as https://github.com/cowprotocol/cowswap/issues/3977 suggest. Although as it is now is ok IMO

@shoom3301 should we kill the `COMING SOON` messages? or we leave this for the launch?

alfetopito commented on apps/widget-configurator/src/app/configurator/hooks/useWidgetParamsAndSettings.ts:
```diff
@@ -59,6 +61,14 @@ export function useWidgetParams(configuratorState: ConfiguratorState, isDappMode
       hideConnectButton: isDappMode,
+
+      partnerFee:
+        partnerFeeBps && partnerFeeBps > 0
```
Is the first part needed?

`undefined > 0 === false`

shoom3301 replied:
Fixed

alfetopito reviewed (approved):
Minor nitpicks

shoom3301 commented:
> > should Recipient field should be mandaroty when a user fills in the Partner fee field?
> 
> Yes! In the configurator I don't care too much, we could even potentially hide it as #3977 suggest. Although as it is now is ok IMO
> 
> @shoom3301 should we kill the `COMING SOON` messages? or we leave this for the launch?

@anxolin good point, I will remove them in another PR

shoom3301 commented on apps/cowswap-frontend/src/modules/injectedWidget/utils/validatePartnerFee.ts:
```diff
@@ -0,0 +1,32 @@
+  const { bps, recipient } = input
+
+  const feeTooHighError =
+    bps > PARTNER_FEE_MAX_BPS ? `Partner fee can not be more than ${PARTNER_FEE_MAX_BPS / 100}%!` : undefined
```
Fixed

shoom3301 commented on apps/cowswap-frontend/src/modules/injectedWidget/utils/validateWidgetParams.ts:
```diff
@@ -0,0 +1,29 @@
+
+    if (!validation) return acc
+
+    const errors = validation(params[key] as never)
```
To be honest, I couldn't beat typescript here :( 
I will be happy to improve it, but have no Idea how to do that.

<img width="1093" alt="image" src="https://github.com/cowprotocol/cowswap/assets/7122625/af7e3f42-1cad-4f71-b9c5-d103d1ba99c2">
